src/Npgsql/MultiplexingDataSource.cs

```diff
@@ -170,25 +170,43 @@ async Task MultiplexingWriteLoop()
                     // There were no idle connectors and we're at max capacity, so we can't open a new one.
                     // Enter over-capacity mode - find an unlocked connector with the least currently in-flight
                     // commands and sent on it, even though there are already pending commands.
+                    // We skip connectors that are either waiting for IO or those that are reserved by pool waiters for the purpose of exclusive use.
                     var minInFlight = int.MaxValue;
                     foreach (var c in Connectors)
                     {
-                        if (c?.MultiplexAsyncWritingLock == 0 && c.CommandsInFlightCount < minInFlight)
+                        if (c?.MultiplexAsyncWritingLock == 0 && c.CommandsInFlightCount < minInFlight && c.ReservedForExclusiveUse == 0)
                         {
                             minInFlight = c.CommandsInFlightCount;
-                            connector = c;
+                            // We either have multiplexing work or we're idle, if neither is the case we must be doing 'exclusive use' work (transaction/copy/etc).
+                            // Don't store these connectors, if we end up storing no connector we know all are busy with non-multiplexing work.
+                            if (minInFlight > 0 || c.Connection is null)
```

Comment:
I'm fairly certain this is the correct way to check for all the non-exclusive use cases. But given all the concepts we have I might have missed a subtlety here.